### PR TITLE
OpenTelemetry Logging

### PR DESCRIPTION
**Pull request by wyattjoh**
This PR serves to add support for OpenTelemetry logging via patching the Console API. This is done with some new configuration options that ensure that this feature is opt-in:

```ts
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  experimental: {
    trace: {
      serviceName: 'datadog-otel-poc',
      logging: {
        // When true, this enables automatic patching of the Console API to send
        // messages over to the configured OpenTelemetry endpoint.
        patchConsole: true,
        // (Optional) The URL for which the log lines will be sent to via JSON
        // encoded HTTP-POST.
        // Defaults to `http://localhost:4318/v1/logs`.
        // endpoint: 'http://localhost:4318/v1/logs',
        // (Optional) The maximum number of log items to collect before
        // flushing. When logs are being emitted, if this maximum is reached
        // before the logs are scheduled to be sent, it will be sent
        // immediately. Defaults to `100`.
        // maxBatchSize: 100,
      },
    },
  },
}
```

When `experimental.trace.logging.httpTransportURL` is provided, it will automatically patch the Console API to send log entries formatted in the [OpenTelemetry Logging format](https://opentelemetry.io/docs/reference/specification/logs/data-model/).

For users with more sophisticated logging requirements, they should instead rely on external logging libraries that already offer OpenTelemetry support.

### CODE REVIEW

**Comment by vvo** [comment]
@wyattjoh Thank you for the update.

1. Is there any example in your PR that is using the new flag? Something along the lines of console.log('message'); in an API route that gets sent to the collector
2. Batching: I remember issues in Serverless Functions (lambda) about trying to send data AFTER the response was sent to the user. Because the lambda would be stopped already. Will batching still works here?
3. Were you able to test that with any provider already?

In general: is there a way I can test all of that myself? Even in the PR. Thanks!

File: packages/next/server/config-shared.ts [hunk 1]
```diff
@@ -184,6 +184,25 @@ export interface ExperimentalConfig {
...
+
+    /**
+     * Patch the Console API to automatically send log lines emitted using it to
+     * the OpenTelemetry endpoint via JSON-HTTP. This isn't required if you're
+     * using a third party logging library.
+     */
+    logging?: {
+      /**
+       * The maximum number of log items to collect before flushing. When logs are
+       * being emitted, if this maximum is reached before the logs are scheduled to
```
Comment (vvo) [inline packages/next/server/config-shared.ts:202]: `before the logs are scheduled to be sent`

What's the schedule? Tick based or timeout based?

Comment (wyattjoh) [reply packages/next/server/config-shared.ts:202]: Great question! Right now the scheduler is tick based if `process.nextTick` is available, or immediate if not. I can clarify in the comment.

**Comment by wyattjoh** [comment]
Thanks for the feedback @vvo!

> 1. Is there any example in your PR that is using the new flag? Something along the lines of console.log('message'); in an API route that gets sent to the collector

I can add an example that showcases the new flag.

> 2. Batching: I remember issues in Serverless Functions (lambda) about trying to send data AFTER the response was sent to the user. Because the lambda would be stopped already. Will batching still works here?

This is why I added support for configurable schedulers to the collector. Right now, it switches to the `NoopScheduler` which does no batching and sends the log lines immediately based on the availability of `process.nextTick`. If there are other signals indicating that we are in an environment (like lambda) that would benefit from the `NoopScheduler`, I think we can adjust the heuristic.

> 3. Were you able to test that with any provider already?

So far, I've tested it with a custom server [written here](https://gist.github.com/wyattjoh/1107dae880d66c2963b4ed4366f5bc31) that collects the JSON and writes it to a file. To run, you should compile this program with `go build -o ./postbin .` and run with `./postbin`. Options can be seen by running `./postbin -h`.

I plan on validating it interacts with the collector soon, but wanted to verify that the rest of the implementation is sound before that step.

**Comment by vvo** [comment]
> I plan on validating it interacts with the collector soon, but wanted to verify that the rest of the implementation is sound before that step.

As for the Next.js implementation: I am not the best to judge it.
As for the Next.js API we provide to developers:
- experimental.trace.logging.httpTransportURL => experimental.trace.logging.endpoint? I am not 100% sure we will be doing HTTP transport all the time. So endpoint + the right documentation may be enough? Only a proposal
- console.log/warn/error being wrapped => As you said it would be awesome to be able to test further. For example, if we could run a local example in this PR that would send logs to Datadog (with-trace-logging). Possibly using the real OTEL collector like the winston example? This way I could give you even more feedback on how this is working well when connected to real logging services.

**Comment by wyattjoh** [comment]
> experimental.trace.logging.httpTransportURL => experimental.trace.logging.endpoint? I am not 100% sure we will be doing HTTP transport all the time. So endpoint + the right documentation may be enough?

Given that, I'd actually suggest we keep it as `httpTransportURL`. If in the future we opt to use a different mechanism, we can introduce new configuration paramters. I've adapted the procedure for the URL detection to match that of what's [used in the `@opentelemetry/exporter-trace-otlp-http` package](https://github.com/open-telemetry/opentelemetry-js/blob/369b07e1c7483556fa9da952ba27c2e2828ecefb/experimental/packages/exporter-trace-otlp-http/src/platform/node/OTLPTraceExporter.ts#L54-L65), so it now can depend on environment variables. The `httpTransportURL` might be deprecated in the future if we feel it's dupliciative or confusing.


>  ... if we could run a local example in this PR that would send logs to Datadog...

I've added an example (with documentation) in the `examples/with-open-telemetry` example. I've been able to validate that logs are being sent to Datadog with this implementation.

**Pull request closed without merging**